instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt

```diff
@@ -511,6 +512,293 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
+                .baseUrl(mockWebServerRule.baseUrl)
+                .coordinatesList(mockRoute.routeWaypoints).build()
+        ).getSuccessfulResultOrThrowException().routes
+        mapboxNavigation.setNavigationRoutes(routes)
+
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+
+        val rerouteResult = mapboxNavigation.routesUpdates().filter {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+        }.first()
+        assertEquals(0, rerouteResult.initialLegIndex)
+    }
+
+    @Test
+    fun reroute_on_multieg_route_without_alternatives() = sdkTest {
+        val mapboxNavigation = createMapboxNavigation()
+        val mockRoute = RoutesProvider.dc_very_short_two_legs(activity)
+        val secondLegLocation = mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = mockRoute.routeWaypoints[1].latitude()
+            longitude = mockRoute.routeWaypoints[1].longitude()
+        }
+        val offRouteLocationUpdate = mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = secondLegLocation.latitude + 0.002
+            longitude = secondLegLocation.longitude
+        }
+
+        mockWebServerRule.requestHandlers.addAll(mockRoute.mockRequestHandlers)
+        mockWebServerRule.requestHandlers.add(
+            MockDirectionsRequestHandler(
+                profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+                jsonResponse = readRawFileText(
+                    activity,
+                    R.raw.reroute_response_dc_very_short_two_legs
+                ),
+                expectedCoordinates = listOf(
+                    Point.fromLngLat(
+                        offRouteLocationUpdate.longitude,
+                        offRouteLocationUpdate.latitude
+                    ),
+                    mockRoute.routeWaypoints.last()
+                ),
+                relaxedExpectedCoordinates = true
+            )
+        )
+
+        mapboxNavigation.startTripSession()
+        val routes = mapboxNavigation.requestRoutes(
+            RouteOptions.builder()
+                .applyDefaultNavigationOptions()
+                .applyLanguageAndVoiceUnitOptions(activity)
+                .baseUrl(mockWebServerRule.baseUrl)
+                .coordinatesList(mockRoute.routeWaypoints).build()
+        ).getSuccessfulResultOrThrowException().routes
+        mapboxNavigation.setNavigationRoutes(routes)
+
+        mapboxNavigation.routeProgressUpdates().first()
+        mapboxNavigation.navigateNextRouteLeg()
+        mockLocationReplayerRule.loopUpdate(secondLegLocation, 100)
+        mapboxNavigation.routeProgressUpdates().filter { it.currentLegProgress?.legIndex == 1 }
+            .first()
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+
+        val rerouteResult = mapboxNavigation.routesUpdates().filter {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+        }.first()
```

Comment:
How do you choose how many location updates need to be generated? As I see it's 100 for arrival and 5 for reroute.

What do you think about introducing a function, which updates location until some condition is met?

```suggestion
        stayOnPositionUntill(secondLegLocation) {
            mapboxNavigation.routeProgressUpdates()
                .filter { it.currentLegProgress?.legIndex == 1 }
                .first()
        }
       
       val rerouteResult = stayOnPositionUntill(offRouteLocationUpdate) {
               mapboxNavigation.routesUpdates().filter {
                   it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
                }.first()
       }
```

Reply:
The number of times doesn't really matter, because a new `loopUpdate` will stop the replayer and clear all the events. We just have to make multiple updates in order for NN to switch to this location.
But I like your suggestion.